### PR TITLE
Initial authorization on workspace and workspace-templates

### LINKED ISSUE #96: Enforce auth on POST /workspaces API endpoint
## Description

As a TRE developer
I want the user authorization integrated within the API
So that we can have a consistent way of authorizing callers.

Also see [user roles](/docs/user-roles.md) for additional context.
Predecessor: #95 

## Acceptance criteria

- [x] Only TRE Administrators should be able to create workspaces.
- [x] Validate auth token for issuer and audience.
- [x] Validate incoming user role claim. Roles contains 'TRE Admin'
- [x] Developer documentation describing how to require an user role to access an API endpoint.
- [x] Developer documentation on how to call the protected API
- [x] No hard dependency on Azure AD SDK - Dependency on OpenId Connect or create abstraction to allow for other OpenId Connect services.

### LINKED ISSUE #310: [Task] Implement auth token validation for API
Story: #305

**Description**
Sets up the basics for auth token validation for the management API

### LINKED ISSUE #311: [Task] Enforce AUTH for GET /workspaceTemplates
Story: #305 

**Description**
Only Azure TRE admins should have access to GET /workspaceTemplates

### LINKED ISSUE #312: [Task] Enforce AUTH for POST /workspaceTemplates
Story: #305 

**Description**
Only Azure TRE admins should have access to POST /workspaceTemplates

### LINKED ISSUE #313: [Task] Enforce AUTH for GET /workspaceTemplates/{template_name}
Story: #305 

**Description**
Only Azure TRE admins should have access to GET /workspaceTemplates/{template_name}

### LINKED ISSUE #314: [Task] Configure swagger UI to use OpenID
Story: #305

It should be easy to authenticate via Azure AD when using the Swagger UI. Configure the [Swagger UI to authenticate](https://swagger.io/docs/specification/authentication/openid-connect-discovery/) against Azure AD

### PR DESCRIPTION
**Pull request by askew**
# PR for issue #310
(ignore branch name!)

## What is being addressed

This PR adds the requirement for a JWT Bearer token to authorize calls to `/workspaces/*` and `/workspace-templates/*`

This implementation supports the Azure AD v2 endpoints.

Notes:
- The token is validated and admin role is validated but access to user workspaces is not implemented in this PR
- The necessary API ENV vars will be set in a separate PR along with a script to create the necessary AAD Applications

Closes #96 - Enforce AUTH on POST /workspaces - only TRE Admin
Closes #310 - Implement AUTH token validation for API
Closes #311 - Enforce AUTH on GET /workspaceTemplates - only TRE Admin
Closes #312 - Enforce AUTH on POST /workspaceTemplates - only TRE Admin
Closes #313 - Enforce AUTH on GET /workspaceTemplates/{template_name} - only TRE Admin
Closes #314 - Configure swagger to use OpenID AAD AUTH

This PR also sets the required environment variables in the Action that deploys to the Dev environment.

### CODE REVIEW

File: management_api_app/services/authentication.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    @property
+    def email(self):
+        return self._get_claim('email')
+
+    @property
+    def roles(self):
+        return self._get_claim('roles')
+
+    def _get_claim(self, claim: str):
```
Comment (sachinkundu) [inline management_api_app/services/authentication.py:30]: _claims is a dict an dict has a .get(blah, default). Instead of checking if it exists you should use that for more pythonic idiom

File: management_api_app/services/authentication.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
+from fastapi import Request, Depends, HTTPException, status
+from fastapi.security import OAuth2AuthorizationCodeBearer, HTTPBearer
+from typing import List, Optional
+from core import config
+import base64
+import jwt
+import requests
+import rsa
+
+class User:
```
Comment (sachinkundu) [inline management_api_app/services/authentication.py:10]: Also is User a wrapper over a dict? Why not use User(dict) as delegate everything to it?

File: management_api_app/services/authentication.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    _jwt_keys: dict = {}
+
+    def __init__(
+        self,
+        instance: str,
+        tenant: str,
+        auto_error: bool = True,
+    ):
+        super(AzureADAuthorization, self).__init__(
+            authorizationUrl = "{instance}/{tenant}/oauth2/v2.0/authorize".format(tenant=tenant,instance=instance),
```
Comment (sachinkundu) [inline management_api_app/services/authentication.py:44]: consider using f strings since they are so much more readable and we have recent enough python in use

File: management_api_app/services/authentication.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+        credentials_exception = HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Could not validate credentials",
+            headers={"WWW-Authenticate": "Bearer"},
+        )
+        try:
+            decoded_token = self._validate_token(token)
+
+            if 'roles' not in decoded_token or len(decoded_token['roles']) == 0:
```
Comment (sachinkundu) [inline management_api_app/services/authentication.py:65]: consider extracting the predicate into has_roles one line function which makes it easier for reading.

File: management_api_app/services/authentication.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not assigned to any role.")
+
+            return User(decoded_token)
+        except:
+            raise credentials_exception
```
Comment (sachinkundu) [inline management_api_app/services/authentication.py:70]: logging credentials exception is a must have

File: management_api_app/services/authentication.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    def _validate_token(self, token: str):
+        key_id = self._get_key_id(token)
+
+        key = self._get_token_key(key_id)
+
+        decoded  = jwt.decode(token, key, verify=True, algorithms=['RS256'], audience=config.API_AUDIENCE)
+
+        return decoded
+
+    def _get_key_id(self, token: str):
```
Comment (sachinkundu) [inline management_api_app/services/authentication.py:82]: consider making utility functions which don't need self be marked as staticmethod for clarity

File: management_api_app/services/authentication.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    # The base64 encoded keys are not always correctly padded, so pad with the right number of =
+    def _ensure_b64padding(self, key: str):
+        key = key.encode('utf-8')
+        missing_padding = len(key) % 4
+        for _ in range(missing_padding):
+            key = key + b'='
+        return key
+
+    # Rather tha use PyJWKClient.get_signing_key_from_jwt every time, we'll get all the keys from AAD and cache them.
+    def _get_token_key(self, key_id: str):
```
Comment (sachinkundu) [inline management_api_app/services/authentication.py:97]: This function is a bit hard to follow with lots of conditions. would it be possible to do method extract refactoring so that the intent become more clear. 

Also consider using functools lru_cache if it makes sense and then you get automatic memoization instead of doing it manually.

**Comment by sachinkundu** [comment]
I was going to comment on using Depends instead of Middleware for auth and then I read that in FastAPI Depends is the canonical method. Weird I say but we all have opinions :)

I am concerned that with auth in working locally is going to become challenging. Is it possible to somehow create non auth FastAPI app in debug mode or we can make LOCAL mode.

**Comment by TessFerrandez** [comment]
> I was going to comment on using Depends instead of Middleware for auth and then I read that in FastAPI Depends is the canonical method. Weird I say but we all have opinions :)
> 
> I am concerned that with auth in, working locally is going to become challenging. Is it possible to somehow create non auth FastAPI app in debug mode or we can make LOCAL mode.

As we discussed in the daily - we won't implement a debug route for security purposes -- but, for temporary debugging purposes you can override the dependencies (as in the app test fixture)

File: management_api_app/api/routes/workspace_templates.py [hunk 8]
```diff
@@ -3,28 +3,35 @@
...
-    WorkspaceTemplateNamesInList,
-    WorkspaceTemplateInCreate,
-    WorkspaceTemplateInResponse
-)
+from models.schemas.workspace_template import (WorkspaceTemplateNamesInList, WorkspaceTemplateInCreate, WorkspaceTemplateInResponse)
 from resources import strings
+from services.authentication import User
 
 
 router = APIRouter()
```
[the remark this answers is not in the document]
Comment (TessFerrandez) [reply management_api_app/api/routes/workspace_templates.py:15]: Let's discuss - but there are no dependencies that will run for all -- for example health/status will not have auth.
Some of the others will need admin rights and some will need specific rights based on what workspace they are accessing, so what dependences do you see would fit here?

[the remark this answers is not in the document]
Comment (TessFerrandez) [reply management_api_app/api/routes/workspace_templates.py]: Fixed per our discussion offline

File: management_api_app/tests/test_api/conftest.py [hunk 9]
```diff
@@ -16,10 +16,21 @@ def no_database():
                     yield
 
 
+def override_get_user():
```
Comment (sachinkundu) [inline management_api_app/tests/test_api/conftest.py:19]: are you then running all tests with admin role? 

We should specifically write atleast one test for each route that privileged route should not execute using non privileged user.

**Review by sachinkundu — APPROVED** [review]
consider writing non privileged tests

File: management_api_app/tests/test_api/conftest.py (hunk 9 above)
Comment (TessFerrandez) [reply management_api_app/tests/test_api/conftest.py:19]: Created #340 for this purpose
